### ISSUE TITLE
Adam (fused=True) issues

### ISSUE THREAD

**Issue by vince62s** [pytorch/pytorch#90752]
### 🐛 Describe the bug

I am having serious doubt on the 1.13 Adam(fused=true) implementation.
First it should be faster than fused=False, in our case it is not, but also the numerical seems off.

Adam(fused=False)
[2022-12-09 12:21:37,894 INFO] Start training loop and validate every 10000 steps...
[2022-12-09 12:23:49,851 INFO] Step 100/100000; acc: 16.1; ppl: 6103.2; xent: 8.7; lr: 0.00002; sents: 169920; bsz: 8605/10659/340; 32607/40388 tok/s; 132 sec;
[2022-12-09 12:25:22,693 INFO] Step 200/100000; acc: 21.4; ppl: 957.8; xent: 6.9; lr: 0.00005; sents: 153056; bsz: 8588/10584/306; 46250/57000 tok/s; 225 sec;
[2022-12-09 12:26:55,347 INFO] Step 300/100000; acc: 25.8; ppl: 414.5; xent: 6.0; lr: 0.00007; sents: 139424; bsz: 8796/10695/279; 47467/57717 tok/s; 317 sec;

Adam(fused=True)
[2022-12-09 12:29:46,843 INFO] Step 100/100000; acc: 15.9; ppl: 7363.9; xent: 8.9; lr: 0.00002; sents: 169920; bsz: 8605/10659/340; 32465/40212 tok/s; 133 sec;
[2022-12-09 12:31:19,458 INFO] Step 200/100000; acc: 18.1; ppl: 1606.6; xent: 7.4; lr: 0.00005; sents: 153056; bsz: 8588/10584/306; 46363/57140 tok/s; 225 sec;
[2022-12-09 12:32:53,634 INFO] Step 300/100000; acc: 22.1; ppl: 702.1; xent: 6.6; lr: 0.00007; sents: 139424; bsz: 8796/10695/279; 46700/56784 tok/s; 319 sec;

Fusedadam apex O2
[2022-12-09 12:35:51,087 INFO] Step 100/100000; acc: 16.1; ppl: 6136.2; xent: 8.7; lr: 0.00002; sents: 169920; bsz: 8605/10659/340; 36301/44964 tok/s; 119 sec;
[2022-12-09 12:37:11,869 INFO] Step 200/100000; acc: 21.4; ppl: 959.9; xent: 6.9; lr: 0.00005; sents: 153056; bsz: 8588/10584/306; 53155/65510 tok/s; 199 sec;
[2022-12-09 12:38:33,601 INFO] Step 300/100000; acc: 25.8; ppl: 415.1; xent: 6.0; lr: 0.00007; sents: 139424; bsz: 8796/10695/279; 53810/65430 tok/s; 281 sec;

Fusedadam apex old legacy code
[2022-12-09 12:42:04,245 INFO] Step 100/100000; acc: 16.1; ppl: 6136.2; xent: 8.7; lr: 0.00002; sents: 169920; bsz: 8605/10659/340; 37141/46004 tok/s; 116 sec;
[2022-12-09 12:43:21,080 INFO] Step 200/100000; acc: 21.4; ppl: 959.9; xent: 6.9; lr: 0.00005; sents: 153056; bsz: 8588/10584/306; 55886/68875 tok/s; 193 sec;
[2022-12-09 12:44:41,898 INFO] Step 300/100000; acc: 25.8; ppl: 415.1; xent: 6.0; lr: 0.00007; sents: 139424; bsz: 8796/10695/279; 54419/66169 tok/s; 274 sec;

As you can see the second one (Adam(fused=True) is not faster, clearly much slower than the Apex implementations, and accuracy, ppl, loss are off compared to the 3 others.

### Versions

pytorch 1.13
Ubuntu 20.04
OpentNMT-py 3.0.1

**Comment by ngimel** [comment]
cc @crcrpar, @ptrblck

**Comment by albanD** [comment]
Thanks for the report, could you also share the number, shape and dtype of your parameters please?

**Comment by vince62s** [comment]
Transformer Large (so about 200M parameters) fp16.

**Comment by vince62s** [comment]
@albanD any clue?

**Comment by albanD** [comment]
Would need to investigate.
cc @crcrpar who is working on these fused optimizers. Does that ring a bell?

**Comment by ngimel** [comment]
also cc @janeyx99

**Comment by crcrpar** [comment]
I'm working on it

**Comment by crcrpar** [comment]
It looks like `Adam(..., fused=True)` in OpenNMT-Py unscales gradients twice when `torch.cuda.amp.GradScaler` is employed.
First unscaling happens https://github.com/OpenNMT/OpenNMT-py/blob/3c4f6bc808d0f35834e2dd4bd7ea746b339921b1/onmt/utils/optimizers.py#L371-L372 and the second happens in the fused kernel https://github.com/pytorch/pytorch/blob/63115b70f03da883a4d2864245eadd2bd5879c1e/aten/src/ATen/native/cuda/fused_adam_utils.cuh#L43-L45.
We could take `OptState` into account to decide whether or not to register `found_inf` and `grad_scale` or pass `GradScaler` itself in https://github.com/pytorch/pytorch/blob/63115b70f03da883a4d2864245eadd2bd5879c1e/torch/cuda/amp/grad_scaler.py#L329 to avoid this hiccup.
#92617.

@vince62s would you know why OpenNMT-Py manually calling `unscale_`?
from the docs of https://pytorch.org/docs/stable/amp.html#torch.cuda.amp.GradScaler.unscale_, it doesn't look necessary

**Comment by vince62s** [comment]
good question why manually, but I am not sure this is happening twice because a few lines after you can see another if statement that does not step the same way.

@Zenglinxiao @francoishernandez might know better why this was done manually.

**Comment by crcrpar** [comment]
I took a profile of fused=True and fused=False and the captures are as follow (top: fused, bottom: normal).
specifically speaking about the duration of `optimizer.step`, at least the fused took shorter (cuda 6.068ms, cpu 3.201ms) than the other (cuda 28.757ms, cpu 28.214ms) on A100 & cuda12, 

`Adam(..., fused=True)`
![image](https://user-images.githubusercontent.com/16191443/216556601-8784518f-d6e1-42b3-8106-9e554519b3bc.png)

vanilla Adam
![image](https://user-images.githubusercontent.com/16191443/216556701-6eeb997b-67ca-423d-9804-d7c96dd77421.png)
I checked

**Comment by vince62s** [comment]
on our codebase ?

**Comment by vince62s** [comment]
Just letting you know the PR fixed the numerical inaccuracy but the performance remains the same.

**Comment by janeyx99** [comment]
@crcrpar I'm doing some old notification cleanup--is there anything actionable we can do here?
